### ISSUE TITLE
Auto sync

### ISSUE THREAD

**Issue by erm3nda** [pgptool/pgptool#208]
I've readed:
https://github.com/pgptool/pgptool/issues/117
https://github.com/pgptool/pgptool/issues/131
https://github.com/pgptool/pgptool/issues/155

I am not sure if I missed it, but i would like to have something like "encrypt back on modify" in an automatic fashion. I don't expect to go to another window then click anything at all.

My goal:
Keepass actually has the "reload if changed" option, that combined with "autosave" or "save on edit" reduces a lot the time lapse between changes so it's not my real intention to be making a lot of changes in real-time. Having sync-solving feature to choose which one in case that both changes it's a plus. 

But the BIG part missing is a sync feature, that let's you sync "decrypted to encrypted" and reverse, when changes in any side.

Case this is not even planned, i would digg a bit for "just sync" programs.
I tried Cryptomator for such thing, but the Android client stopped me, and Encdroid is not being developed anymore and needs a lot of permissions and FUSE to make it really work as the desktop client does.

That's why my research falled back to use PGP in a "file to file" basis.

Regards.

**Comment by skarpushin** [comment]
hi @erm3nda 
Ok, so context of this conversation is feature request for "Automatic Encrypt Back invocation when PGP Tool detects changes of the decrypted file".

Although I think I understand where you're coming from with this request, I'm concerned a little bit that this feature will reward having sensitive information in unencrypted state on your disk. WDYT about this?

Also, I have couple questions, how do you see this affects User Interface.
Let's pretend this feature is implemented. How do you see it is working in following cases:

1. How this feature will be turned on/off - per file (i.e. checkbox in decryption window) or for all files?
1. How successfull automatical encryption is communicated in the user interface / to the user? 
1. What happens if PGP Tool detected file change, but cannot read it because it is blocked by other application?
1. What happens if PGP Tool started encryption, but decrypted file got changed in a meantime?
1. What happens if PGP Tool tried to save newly encrypted file, but latter happened to be blocked by other application?
1. What happens if some other type of failure happened somewhere in this process - how error is communicated to the user and what actions PGP Tool will perform?

And another thing: am I understanding correctly that you're aware of the "Encrypt Back All" feature?
In general it allows you to encrypt back all files in simple 4 clicks:
1. Click on the PGP Tool in task bar
2. Click on "Encrypt back all" button
3. Click to confirm operation
4. Click to confirm operation results

**Comment by erm3nda** [comment]
Thank you for adding the feature, mi fail.

> am I understanding correctly that you're aware of the "Encrypt Back All" feature?
Yes.

I am seeing new things about my request at the time of writting. You made good questions. **Let me a bit more time to analyze and expose it properly.**

Be aware that i am asking from the only perspective of Keepassxc program, so anything i say may not apply other's usage.

In the other hand, having such data unencryted doesn't affect Keepassxc database wich is already a 256bit encrypted file. This mean that file in the could will be double encrypted one, so it's safe to mantain it for "check file date" purposes instead of make a hard to follow saving-erasing-saving game.

Regards.

**Comment by erm3nda** [comment]
> How this feature will be turned on/off - per file (i.e. checkbox in decryption window) or for all files?
Ideally to be both a Global option and per item option. Just trigering the "save back" in automaticly basis. It will be exactly as if I where monkey clicking it manually.

> How successfull automatical encryption is communicated in the user interface / to the user?
I think you're referring on how will i Know that the item is "synced" or somehow. A little "green tick" that says that the file has been "written back" at least 1-5-10 secods ago" could be enough.

> What happens if PGP Tool detected file change, but cannot read it because it is blocked by other application?
In the particular case of Keepassxc, the file is adopted and used in RAM. Means that i could still work with my current database even if the file disspears, and put it back just "saving", wich can be done automaticly on_changes or on save_button.
In case of other app's that are being blocked, just as the user to "accept or reject". I do understand that this could be a killing task to solve, but happens everyday with any syncing "fashioned" program. Unless you have the hability of merge file contents you're forced to accept, reject or stop.

> What happens if PGP Tool started encryption, but decrypted file got changed in a meantime?
Everything should stay away once encryption/decryption happends. The "writting back" feature should know the encryption_process_status and wait untill is done. Then try. 

> What happens if PGP Tool tried to save newly encrypted file, but latter happened to be blocked by other application?
I don't think that PGP Tool makes magic. If there's a problem with that is a problem of the implementation of those sides. My real petition is "to commit changes back like if i where pushing the button manually". Nothing can stop me to push that button like a monkey. Im just asking for automate that, regardless of the problems that this will put on the road.

> What happens if some other type of failure happened somewhere in this process - how error is communicated to the user and what actions PGP Tool will perform?
I don't really get this question. If some kind of big failure happends, just stop trying to sycn that particular file untill the user solves it. A desktop notification could solve the "notification" issue.

> And another thing: am I understanding correctly that you're aware of the "Encrypt Back All" feature?
> In general it allows you to encrypt back all files in 4 simple clicks:
> Click on the PGP Tool in task bar
> Click on "Encrypt back all" button
> Click to confirm operation
> Click to confirm operation results

Totally yes, I just want a monkey triggering feature that does that for me.
Regarding the question on how to implement it, globally or per-file, now I think that a little opt-in for sycing per file could solve most of the problems. This way i can put that to work my the files i do need (especially little database files) rather than try to make magic with bigger files that could "break" the whole PGP Tool purpose.

Excuse my for the typos, my English is a bit shitty.
Thank you so much. If you have more questions of want to polish those feel free to ask me anytime.

Regards.

**Comment by erm3nda** [comment]
Regarding the exact case of Keepassxc.

I open a file with PGP Tool. It's getting opened but not blocking by Keepassxc.
Keepass will save back on changes. PGP Tool will se the change, then encrypt back. This file will be on a synced folder, letting me to get an updated copy of the changes.
Once any side changes the file, it will be synced. The Keepassxc program does look for file changes by itself. If the ciphered file does change it will be "reloaded" by Keepassxc with no action for the user.

The whole idea is to have a shared, synced, encrprypted database that is accesible and up-to-date on all my systems (windows computer, linux computer, android phone).

The way to accomplish that is that PGP Tool is able to monkey-click the encript back feature and decrypt again on changes.

Regards.

**Comment by skarpushin** [comment]
@erm3nda I think I totally understand what you want to get with this feature. 
Here is my feedback:
- It is not possible to provide rock-solid operation in such scenario. Especially when there is a noticeable probability of failure due to concurrency issues. 
- Also, when it comes to security of sensitive data, I'm concerned about application doing something under the hood with this data.
- And yes, this feature will require a lot of efforts to implement it the right way (happy scenario is not a problem. Problem is -- graceful handling when something goes wrong - also be mindful about User Experience).

I'll keep this feature request in a backlog and get back to it when I have time.

**Comment by erm3nda** [comment]
Perfect, thank you.

The whole idea was to use common providers (dropbox, drive, etc) to store encrypted data.
I tried in first time Cryptomator, but the phone client is not free and I can't afford right now any kind of payment for something that is just for my personal use.

For now, I'll just stay using a webdav provider that I trust enough. As I said, databases are encrypted by themselves.

Feel free to ask me back when you see this again. I don't know to code Java, but I'll check or test this feature if you end up adding something.

Regards.
Closing.
